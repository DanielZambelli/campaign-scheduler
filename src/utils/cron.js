const CronJob = require('cron').CronJob

const cron = (cronStr, callback, autoStart = true) => {
  return new CronJob(cronStr, callback ,null, autoStart)
}

const cronOnce = (cronStr, callback, autoStart = true) => {
  const job = cron(cronStr, async () => {
    await callback()
    job.stop()
  }, autoStart)
  return job
}

module.exports = {cron,cronOnce}
