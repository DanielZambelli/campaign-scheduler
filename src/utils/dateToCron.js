const dateToCron = (dateString) => {
  const date = new Date(dateString)
  const seconds = date.getSeconds()
  const minutes = date.getMinutes()
  const hours = date.getHours()
  const days = date.getDate()
  const months = date.getMonth()
  const dayOfWeek = date.getDay()
  return `${seconds} ${minutes} ${hours} ${days} ${months} ${dayOfWeek}`
}

module.exports = {dateToCron}
