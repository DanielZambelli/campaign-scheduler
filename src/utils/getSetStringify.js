const getSetStringify = (field) => {
  return {
    get(){
      const value = this.getDataValue(field)
      return value ? JSON.parse(value) : value
    },
    set(value){
      if(value) value = JSON.stringify(value)
      this.setDataValue(field, value)
    }
  }
}

module.exports = {
  getSetStringify
}
