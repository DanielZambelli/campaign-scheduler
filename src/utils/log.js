const util = require('util')
const log = (object) => console.log(util.inspect(object, false, null, true))

module.exports = {log}
