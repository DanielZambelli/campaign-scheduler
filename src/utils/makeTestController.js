const { CampaignScheduler: Controller } = require('../controller/index')

const makeTestController = (id, callbacks = {}) => {

  const db = process.env.CS_DB ? JSON.parse(process.env.CS_DB) : { "dialect": "sqlite" }
  if(db.dialect==='sqlite') db.storage = `tmp/db_${id}.sqlite`

  db.schema = `test_${id}`
  db.force = true
  db.makeDestroy = true

  return new Controller({ db, callbacks }).init()
}

module.exports = {
  makeTestController
}
