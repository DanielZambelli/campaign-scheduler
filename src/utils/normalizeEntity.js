const normalizeEntity = (entity) => {
  const {createdAt,updatedAt,...entityNormalized} = entity.toJSON()
  return entityNormalized
}

module.exports = {
  normalizeEntity
}
