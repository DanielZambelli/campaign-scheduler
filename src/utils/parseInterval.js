const moment = require('moment')
const {parseIntervalOffset} = require('./parseIntervalOffset')
const {parseIntervalTime} = require('./parseIntervalTime')

const parseInterval = (interval, startDate=undefined) => {

  const {offset=null, time=null} = interval
  startDate = moment(startDate)

  //start from now or offset from previous date
  const date = moment(startDate)

  // offset
  const offsetParsed = parseIntervalOffset(offset)
  if(offsetParsed) date.add(offsetParsed)

  // specific time
  const timeParsed = parseIntervalTime(time)
  if(timeParsed) date.set(timeParsed)

  // incase now has passed the time slot offset till next day
  if(date.isSame(startDate, 'day') && startDate.isAfter(date))
    date.add({ days: 1 })

  return date.toISOString()
}

module.exports = {parseInterval}
