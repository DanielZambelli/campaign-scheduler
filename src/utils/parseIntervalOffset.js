const {getRandomBetween} = require('./getRandomBetween')

const parseIntervalOffset = (offset) => {
  if(!offset) return
  let [a,b] = offset.split('-').map(offset => {
    return offset.trim().split(' ').reduce((map,item) => {
      const value = parseInt(item.match(/\d+/g))
      const unit = item.replace(value,'')
      map[unit] = value
      return map
    }, {})
  })
  if(a && b){
    Object.keys(b).forEach(unit => {
      if(a[unit]){
        const min = Math.min(a[unit], b[unit])
        const max = Math.max(a[unit], b[unit])
        a[unit] = getRandomBetween(min,max)
      }else a[unit] = b[unit]
    })
  }
  return a
}

module.exports = {parseIntervalOffset}
