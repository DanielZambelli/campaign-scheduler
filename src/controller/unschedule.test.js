require('../utils/test')
const {seed} = require('../utils/test/seeds1')
const TEST = 'unschedule'

describe(TEST, () => {

  beforeEach(initCtl(TEST,seed))

  afterEach(destroy)


  it('validates', async () => {
    const res = await ctl.unschedule('UNEXISTING_CAMPAIGN_ID', 'contact', 1).catch(e => e.message)
    expect(res).toEqual('schedule not found')
  })

  it('validates', async () => {
    const res = await ctl.unschedule('UNEXISTING_CAMPAIGN_ID', 'portal', 1).catch(e => e.message)
    expect(res).toEqual('schedule not found')
  })

  it('returns', async () => {
    await ctl.schedule(seeds.campaign1.id, 'contact', 1)
    const res = await ctl.unschedule(seeds.campaign1.id, 'contact', 1).then(normalize)
    expect(res.hash).toEqual('1d939651eef18dc36074572891338df6e4d900e4')
  })

  it('returns', async () => {
    await ctl.schedule(seeds.campaign1.id, 'contact', 1)
    await ctl.schedule(seeds.campaign1.id, 'portal', 1)
    const res = [
      await ctl.unschedule(seeds.campaign1.id, 'contact', 1).then(normalize),
      await ctl.unschedule(seeds.campaign1.id, 'portal', 1).then(normalize),
    ]
    expect(res[0].hash).toEqual('1d939651eef18dc36074572891338df6e4d900e4')
    expect(res[1].hash).toEqual('0bc6011a84ee30e68c7284f0f850164b16945462')
  })

  it('returns', async () => {
    await ctl.schedule(seeds.campaign2.id, 'contact', 1)
    const res = await ctl.unschedule(seeds.campaign2.id, 'contact', 1).then(normalize)
    expect(res.hash).toEqual('6b5fcf1bebec278a80e077cdea3e2b4f9cc506c0')
  })

  it('state', async () => {
    await ctl.schedule(seeds.campaign3.id, 'contact', 1)
    await ctl.unschedule(seeds.campaign3.id, 'contact', 1)
    const res = await getState()
    expect(res.hash).toEqual('2a7c7656d407a04fb6dffb5bc8a9a69f6456c730')
  })

  it('state', async () => {
    await ctl.schedule(seeds.campaign1.id, 'contact', 1)
    await ctl.schedule(seeds.campaign1.id, 'portal', 1)
    await ctl.schedule(seeds.campaign2.id, 'contact', 1)
    await ctl.schedule(seeds.campaign3.id, 'contact', 1)
    await ctl.unschedule(seeds.campaign1.id, 'contact', 1)
    await ctl.unschedule(seeds.campaign1.id, 'portal', 1)
    await ctl.unschedule(seeds.campaign2.id, 'contact', 1)
    await ctl.unschedule(seeds.campaign3.id, 'contact', 1)
    const res = await getState()
    expect(res.hash).toEqual('c7b582f255620e5e369ccb8ee843c6e423c492f6')
  })

  it('reconcile pendings', async () => {
    await ctl.schedule(seeds.campaign3.id, 'contact', 1)
    await ctl.schedule(seeds.campaign3.id, 'contact', 2)
    await ctl.schedule(seeds.campaign3.id, 'contact', 3)
    await ctl.Db.Actions.update({ state: 'completed' }, { where: { subjectId: 1, actionId: 'action1'  } })
    await ctl.Db.Actions.update({ state: 'failed' }, { where: { subjectId: 2, actionId: 'action1'  } })
    await ctl.unschedule(seeds.campaign3.id, 'contact', 1)
    await ctl.unschedule(seeds.campaign3.id, 'contact', 2)
    const res = await getState()
    expect(res.hash).toEqual('0eadde3355fa642ee2865e1d1e9474f1d384bf19')
  })

})
