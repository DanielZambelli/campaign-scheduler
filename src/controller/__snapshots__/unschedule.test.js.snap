// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unschedule reconcile pendings 1`] = `
{
  "hash": "4edd6bb74ddfd75758edd4af5ca7a773ea1f8875",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "failed",
        "subject": "contact#2",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "pending",
        "subject": "contact#3",
      },
      {
        "actionId": "action2",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#3",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1days",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": false,
        "campaignId": "campaign3",
        "subject": "contact#1",
      },
      {
        "active": false,
        "campaignId": "campaign3",
        "subject": "contact#2",
      },
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#3",
      },
    ],
  },
}
`;

exports[`unschedule returns 1`] = `
{
  "hash": "4c0f695b59e731f7fcead77888a45b1bf52cc590",
  "result": {
    "actionsDestroyed": 1,
    "schedule": {
      "active": false,
      "campaignId": "campaign1",
      "createdAt": "a few seconds ago",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
  },
}
`;

exports[`unschedule returns 2`] = `
[
  {
    "hash": "4c0f695b59e731f7fcead77888a45b1bf52cc590",
    "result": {
      "actionsDestroyed": 1,
      "schedule": {
        "active": false,
        "campaignId": "campaign1",
        "createdAt": "a few seconds ago",
        "subject": "contact#1",
        "updatedAt": "a few seconds ago",
      },
    },
  },
  {
    "hash": "0540e16fb309a15e572aa1760027bbffe04e9065",
    "result": {
      "actionsDestroyed": 1,
      "schedule": {
        "active": false,
        "campaignId": "campaign1",
        "createdAt": "a few seconds ago",
        "subject": "portal#1",
        "updatedAt": "a few seconds ago",
      },
    },
  },
]
`;

exports[`unschedule returns 3`] = `
{
  "hash": "e5a3377dd8d722a9d91b233c0948be732eda6825",
  "result": {
    "actionsDestroyed": 0,
    "schedule": {
      "active": false,
      "campaignId": "campaign2",
      "createdAt": "a few seconds ago",
      "subject": "contact#1",
      "updatedAt": "a few seconds ago",
    },
  },
}
`;

exports[`unschedule state 1`] = `
{
  "hash": "89df79f373c9b9a2267c23b8a181f691d3fee344",
  "state": {
    "actions": [],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1days",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": false,
        "campaignId": "campaign3",
        "subject": "contact#1",
      },
    ],
  },
}
`;

exports[`unschedule state 2`] = `
{
  "hash": "781d5433bce6d82ae9292e54ede2981a345acb12",
  "state": {
    "actions": [],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1days",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": false,
        "campaignId": "campaign1",
        "subject": "contact#1",
      },
      {
        "active": false,
        "campaignId": "campaign1",
        "subject": "portal#1",
      },
      {
        "active": false,
        "campaignId": "campaign2",
        "subject": "contact#1",
      },
      {
        "active": false,
        "campaignId": "campaign3",
        "subject": "contact#1",
      },
    ],
  },
}
`;
